enterprise/frontend/src/embedding-sdk/hooks/private/use-init-data.ts

```diff
@@ -36,65 +37,76 @@ const getErrorMessage = (authType: SDKConfigType["authType"]) => {
   return t`Invalid auth type`;
 };
 
+const isJwtAuth = (config: SDKConfigType): config is SdkConfigWithJWT =>
+  config.authType === "jwt" && !!config.jwtProviderUri;
+
+const isApiKeyAuth = (config: SDKConfigType): config is SdkConfigWithApiKey =>
+  config.authType === "apiKey" && !!config.apiKey;
+
 export const useInitData = ({ config }: InitDataLoaderParameters) => {
   const dispatch = useSdkDispatch();
 
   const loginStatus = useSdkSelector(getLoginStatus);
 
-  const [sessionTokenState, setSessionTokenState] =
-    useState<EmbeddingSessionTokenState | null>(null);
-
   useEffect(() => {
     registerVisualizationsOnce();
     dispatch(setLoginStatus({ status: "uninitialized" }));
   }, [dispatch]);
 
-  const jwtProviderUri =
-    config.authType === "jwt" ? config.jwtProviderUri : null;
+  const setupJwtAuth = useCallback(
+    (config: SdkConfigWithJWT) => {
+      api.onBeforeRequest = async () => {
+        const tokenState = await dispatch(
+          getOrRefreshSession(config.jwtProviderUri),
+        );
 
-  useEffect(() => {
-    if (config.authType === "jwt") {
-      const updateToken = () => {
-        const currentState = store.getState();
-        setSessionTokenState(getSessionTokenState(currentState));
+        api.sessionToken = (
+          tokenState.payload as EmbeddingSessionTokenState["token"]
+        )?.id;
       };
-
-      const unsubscribe = store.subscribe(updateToken);
-
-      if (jwtProviderUri) {
-        dispatch(getOrRefreshSession(jwtProviderUri));
-      }
-
-      updateToken();
-
-      return () => unsubscribe();
-    }
-  }, [config.authType, dispatch, jwtProviderUri]);
-
-  useEffect(() => {
-    api.basename = config.metabaseInstanceUrl;
-
-    if (config.authType === "jwt" && config.jwtProviderUri) {
-      api.onBeforeRequest = () =>
-        dispatch(getOrRefreshSession(config.jwtProviderUri));
-      api.sessionToken = sessionTokenState?.token?.id;
       dispatch(setLoginStatus({ status: "initialized" }));
-    } else if (config.authType === "apiKey" && config.apiKey) {
+    },
+    [dispatch],
+  );
+
+  const setupApiKeyAuth = useCallback(
+    (config: SdkConfigWithApiKey) => {
       api.apiKey = config.apiKey;
       dispatch(setLoginStatus({ status: "initialized" }));
-    } else {
-      dispatch(
-        setLoginStatus({
-          status: "error",
-          error: new Error(getErrorMessage(config.authType)),
-        }),
-      );
+    },
+    [dispatch],
+  );
+
+  useEffect(() => {
+    if (loginStatus.status === "uninitialized") {
+      api.basename = config.metabaseInstanceUrl;
+
+      if (isJwtAuth(config)) {
```

Comment:
this is me attempting to fix typescript errors. I'm open to feedback because this seems a little convoluted.

Reply:
I think it's common to use [type predicate](https://www.typescriptlang.org/docs/handbook/advanced-types.html#using-type-predicates) in TypeScript. I think this is totally fine.